### PR TITLE
[tiktok] Add new extractor

### LINKED ISSUE #18108: Site request tiktok.com
### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2018.11.07*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2018.11.07**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x]  [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [ ] Bug report (encountered problems with youtube-dl)
- [x] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

```
[debug] System config: []
[debug] User config: ['-o', '%(autonumber)s-%(title)s_%(upload_date)s.%(ext)s', '--sub-format', 'best', '--write-sub', '--sub-lang', 'en,ja', '--embed-subs', '--netrc']
[debug] Custom config: []
[debug] Command-line args: ['-v', 'https://m.tiktok.com/v/6606727368545406213.html']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.11.07
[debug] Python version 3.7.1 (CPython) - Linux-4.20.0-rc1-mainline-x86_64-with-arch
[debug] exe versions: ffmpeg n4.1, ffprobe n4.1, rtmpdump 2.4
[debug] Proxy map: {}
[generic] 6606727368545406213: Requesting header
WARNING: Falling back on generic information extractor.
[generic] 6606727368545406213: Downloading webpage
[generic] 6606727368545406213: Extracting information
ERROR: Unsupported URL: https://m.tiktok.com/v/6606727368545406213.html
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/youtube_dl/YoutubeDL.py", line 792, in extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python3.7/site-packages/youtube_dl/extractor/common.py", line 508, in extract
    ie_result = self._real_extract(url)
  File "/usr/lib/python3.7/site-packages/youtube_dl/extractor/generic.py", line 3298, in _real_extract
    raise UnsupportedError(url)
youtube_dl.utils.UnsupportedError: Unsupported URL: https://m.tiktok.com/v/6606727368545406213.html
```

---

### If the purpose of this *issue* is a *site support request* please provide all kinds of example URLs support for which should be included (replace following example URLs by **yours**):
- Single video: https://m.tiktok.com/v/6606727368545406213.html
- Single video: https://m.tiktok.com/v/6606468080325364997.html
- Short link format: http://vm.tiktok.com/[A-Z].{5}/

---

### Description of your *issue*, suggested solution and other information

Requesting an extractor for `tiktok.com`/`m.tiktok.com` most videos may only be reached by using the share button from the Android/iOS app. Some videos may be downloaded from the app itself however some have the download option disabled. The site appears to use "jplayer" with html5 video tags and the akamai distribution network. It seems the video src url is loaded via javascript.

### PR DESCRIPTION
**Pull request by Flat**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Provides an extractor for ((v)m).tiktok.com closes #18108

### CODE REVIEW

**Comment by Flat** [comment]
I believe you were referring to optional fields being fatal (using an index which may not exist). The latest commit should remedy that. Let me know if I missed anything else.

**Comment by dstftw** [comment]
Nothing conceptually changed.

**Comment by Flat** [comment]
I hope this is now more in-line with the code conventions.

File: youtube_dl/extractor/tiktok.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        data = self._parse_json(
+            self._search_regex(
+                r'var data = ({.*});', webpage, 'json_string', webpage, 'data'
+            ), video_id)
+
+        title = self._og_search_title(webpage)
+        description = self._og_search_description(webpage)
+
+        video_url = url_or_none(
+            try_get(data, lambda x: x['video']['play_addr']['url_list'][0], compat_str))
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: 1. This is mandatory.
2. All formats must be extracted.

File: youtube_dl/extractor/tiktok.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        uploader = try_get(data, lambda x: x['author']['nickname'], compat_str)
+
+        thumbnail = url_or_none(
+            try_get(
+                data, lambda x: x['video']['cover']['url_list'][0], compat_str))
+
+        ext = str_or_none(
+            urlhandle_detect_ext(
+                self._request_webpage(video_url, video_id, fatal=False)))
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: No.

File: youtube_dl/extractor/tiktok.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        data = self._parse_json(
+            self._search_regex(
+                r'var data = ({.*});', webpage, 'json_string', webpage, 'data'
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: 1. Relax regex.
2. Do not capture empty dict.

File: youtube_dl/extractor/tiktok.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from ..utils import (
+    urlhandle_detect_ext,
+    try_get,
+    compat_str,
+    url_or_none,
+    str_or_none,
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: Alphabetic.

**Comment by Flat** [comment]
I've added all available formats and get the extension in a similar fashion to https://github.com/rg3/youtube-dl/blob/c3bcd206eb031de30179c88ac7acd806a477ceae/youtube_dl/extractor/hearthisat.py#L106-L111 and https://github.com/rg3/youtube-dl/blob/c3bcd206eb031de30179c88ac7acd806a477ceae/youtube_dl/extractor/wdr.py#L75-L78 as the downloads URLs do not provide a file extension.

File: youtube_dl/extractor/tiktok.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        width = int_or_none(try_get(data, lambda x: x['video']['width'], int))
+        height = int_or_none(try_get(data, lambda x: x['video']['height'], int))
+
+        formats = []
+
+        for url in data['video']['play_addr']['url_list']:
+            ext = determine_ext(url)
+            if ext == 'unknown_video':
+                urlh = self._request_webpage(
```
[the remark this answers is not in the document]
Comment (Flat) [reply youtube_dl/extractor/tiktok.py]: Your original comment was "No". Thank you for taking the time to elaborate so I know exactly what to change.

File: youtube_dl/extractor/tiktok.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        data = self._parse_json(
+            self._search_regex(
+                r'var data = ({.+?});', webpage, 'json_string', webpage, 'data'
```
[the remark this answers is not in the document]
Comment (Flat) [reply youtube_dl/extractor/tiktok.py]: Can you please explain what needs to be changed with the regex? Should the white space be optional?

[the remark this answers is not in the document]
Comment (dstftw) [reply youtube_dl/extractor/tiktok.py]: Yes, it should accept any variation of whitespace.

File: youtube_dl/extractor/tiktok.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            for url in url_list[0]:
+                formats.append({
+                    'url': url,
+                    'ext': 'mp4',
+                    'height': height,
+                    'width': width,
+                    'format_note': url_list[1]
+                })
+
+        extract_formats((try_get(data, lambda x: x['video']['play_addr_lowbr']['url_list']), 'Low quality'))
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]:
```
for key, label in (('play_addr_lowbr', 'Low'), ('play_addr', 'Normal'), ('download_addr', 'Download')):
    format_url = url_or_none(try_get(data, lambda x: x['video'][key]['url_list']))
    if not format_url:
        continue
    formats.append(...)
```

File: youtube_dl/extractor/tiktok.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        def extract_formats(url_list):
+            if url_list[0] is None:
+                return
+            for url in url_list[0]:
+                formats.append({
+                    'url': url,
+                    'ext': 'mp4',
+                    'height': height,
+                    'width': width,
+                    'format_note': url_list[1]
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: Should contain `quality` key.

File: youtube_dl/extractor/tiktok.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                })
+
+        extract_formats((try_get(data, lambda x: x['video']['play_addr_lowbr']['url_list']), 'Low quality'))
+        extract_formats((try_get(data, lambda x: x['video']['play_addr']['url_list']), 'Normal quality'))
+        extract_formats((try_get(data, lambda x: x['video']['download_addr']['url_list']), 'Download quality'))
+
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: `_sort_formats`.

File: youtube_dl/extractor/tiktok.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        data = self._parse_json(
+            self._search_regex(
+                r'var\s*data\s*=\s*({.+?});', webpage, 'data'
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: `vardata` makes no sense.

File: youtube_dl/extractor/tiktok.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+class TikTokIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:m\.)?tiktok\.com/v/(?P<id>[0-9]+)'
+    _TEST = {
+        'url': 'https://m.tiktok.com/v/6606727368545406213.html',
+        'info_dict': {
+            'id': '6606727368545406213',
+            'ext': 'mp4',
+            'title': 'Zureeal on TikTok',
+            'thumbnail': 'http://m-p16.akamaized.net/img/tos-maliva-p-0068/5e7a4ec40fb146888fa27aa8d78f86fd~noop.image',
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: `re:`.

File: youtube_dl/extractor/tiktok.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                format_url = url_or_none(format)
+                if not format_url:
+                    continue
+                formats.append({
+                    'url': url,
+                    'ext': 'mp4',
+                    'height': height,
+                    'width': width,
+                    'format_note': label,
+                    'quality': -2 if label == 'Low' else (1 if label == 'Download' else 0)
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: `enumerate` on for range.

File: youtube_dl/extractor/tiktok.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+        formats = []
+
+        for key, label in (('play_addr_lowbr', 'Low'), ('play_addr', 'Normal'), ('download_addr', 'Download')):
+            for format in try_get(data, lambda x: x['video'][key]['url_list']):
+                format_url = url_or_none(format)
+                if not format_url:
+                    continue
+                formats.append({
+                    'url': url,
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: Invalid URL, have you even tested?

Comment (Flat) [reply youtube_dl/extractor/tiktok.py]: Yes I have tested it and it worked. That's why I didn't notice I forgot to update the variable name.
```
python -m youtube_dl https://m.tiktok.com/v/6606727368545406213.html
[TikTok] 6606727368545406213: Downloading webpage
[download] Destination: Zureeal on TikTok-6606727368545406213.mp4
[download] 100% of 68.16KiB in 00:01
```

**Comment by Flat** [comment]
Sorry it took so long to get into a semi-mergable state. Thank you for the reviews and ultimately implementing and improving the extractor. Cheers.

**Pull request closed without merging**